Reorganize and streamline form media type guidance (3.0.4) (#3923)
OAI/OpenAPI-Specification: versions/3.0.4.md

would produce a request body of:

```urlencoded
name=example&icon=iVBORw0KGgoAAAANSUhEUgAAAAIAAAACCAIAAAD91JpzAAAABGdBTUEAALGPC%2FxhBQAAADhlWElmTU0AKgAAAAgAAYdpAAQAAAABAAAAGgAAAAAAAqACAAQAAAABAAAAAqADAAQAAAABAAAAAgAAAADO0J6QAAAAEElEQVQIHWP8zwACTGCSAQANHQEDqtPptQAAAABJRU5ErkJggg%3D%3D
```

> Note that this base64-encoded value had to be futher percent-encoded, replacing `/` with `%2F` and each of two final `=` padding characters with `%3D`.

Reviewer: > replacing `/` with `%2F`

That's why [base64url](https://datatracker.ietf.org/doc/html/rfc4648#section-5) was invented - I wonder why isn't used more often in places where it makes sense.

Author: Yes, it's allowed and used in an example in 3.1.  Oddly, the use of `=` for padding is not changed, so you still end up needing to do percent-encoding :-/

> Using `format: byte` for a multipart field is equivalent to setting `Content-Transfer-Encoding: base64`.

Reviewer 2: I'm confused by this. If `Content-Transfer-Encoding` is deprecated and a multipart field with `format: byte` is equivalent to setting `Content-Transfer-Encoding: base64`, doesn't this imply that a multipart field with `format: byte` is essentially deprecated?

I guess what I'm missing here is why `Content-Transfer-Encoding` was deprecated. It seems like this effectively disallows any encoding of individual fields in a multipart body, which seems undesirable. I'm sure this is just confusion on my part but would appreciate if someone could clear this up for me.

Reviewer 2: I agree that we are obligated to follow the RFCs, which is exactly why I think we need to clearly state that `format: byte` is deprecated for a property in a multipart/form-data body. 

@darrelmiller Would appreciate your thoughts on this.

Author: Deprecated in OAS is not the same as deprecated in RFC7578, and the text I wrote _already_ makes it clear that the _usage_ is deprecated by RFC7578 in this context.

Using `format: byte` for a multipart field is equivalent to setting `Content-Transfer-Encoding: base64`.

> If `format: byte` is used along with setting a different `Content-Transfer-Encoding` value with the `headers` field, the result is undefined.

Reviewer 2: Can we say this instead?
```suggestion
If `format: byte` is specified for a `multipart/form-data` field, the result is undefined.
```

Author: No, because it is completely valid to do this.  "Deprecated" does not mean "forbidden", and the behavior is completely well-defined.  In RFC7578 it is a SHOULD NOT not a MUST NOT, and that RFC was only published in July 2017 – the same month as OAS 3.0.  We support all HTTP APIs, including ones that use behavior that is now deprecated.

The text I wrote in this paragraph provides all of the necessary context for authors to decide whether it is suitable for their use or not.  We wouldn't want to _replace_ this sentence with a `form-data`-specific sentence anyway as it's at least theoretically possible to describe other `multipart` types.
